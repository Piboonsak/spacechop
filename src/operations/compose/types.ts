import * as t from 'runtypes';

const ComposeConfig = t.Record({
  source: t.String,
}).And(t.Partial({
  params: t.Dictionary(t.String, 'string'),
}));

export type ComposeConfig = t.Static<typeof ComposeConfig>;
export default ComposeConfig;
