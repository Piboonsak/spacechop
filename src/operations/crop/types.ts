import * as t from 'runtypes';
import Gravity from './../Gravity';

const CropConfig = t.Partial({
  width: t.Number,
  height: t.Number,
  gravity: Gravity,
}).withConstraint((n) => n && !!(n.width || n.height));

export type CropConfig = t.Static<typeof CropConfig>;
export default CropConfig;
