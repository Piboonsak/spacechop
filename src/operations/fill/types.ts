import * as t from 'runtypes';
import Gravity from './../Gravity';
import PostiveNumber from './../PositiveNumber';

const FillConfig = t.Record({
  width: PostiveNumber,
  height: PostiveNumber,
}).And(t.Partial({
  gravity: Gravity,
}));

export type FillConfig = t.Static<typeof FillConfig>;
export default FillConfig;
