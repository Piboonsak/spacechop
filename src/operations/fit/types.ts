import * as t from 'runtypes';
import Gravity from './../Gravity';

const FitConfig = t.Partial({
  width: t.Number,
  height: t.Number,
  gravity: Gravity,
});


export type FitConfig = t.Static<typeof FitConfig>;
export default FitConfig;
