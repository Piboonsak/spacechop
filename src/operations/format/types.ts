import * as t from 'runtypes';
import Format from './../../types/Format';

const FormatConfig = t.Record({
  type: Format,
});

export type FormatConfig = t.Static<typeof FormatConfig>;
export default FormatConfig;
