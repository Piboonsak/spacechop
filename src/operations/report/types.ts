import * as t from 'runtypes';
import ReportFormat from './../../types/ReportFormat';

const ReportConfig = t.Partial({
  format: ReportFormat,
  compact: t.Boolean,
});

export type ReportConfig = t.Static<typeof ReportConfig>;
export default ReportConfig;
