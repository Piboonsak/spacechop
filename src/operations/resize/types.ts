import * as t from 'runtypes';
import PositiveNumber from '../PositiveNumber';

const ResizeConfig = t.Record({
  width: PositiveNumber,
  height: PositiveNumber,
});

export type ResizeConfig = t.Static<typeof ResizeConfig>;
export default ResizeConfig;
