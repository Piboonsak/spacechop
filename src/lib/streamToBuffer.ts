import { Stream } from 'stream';

export default (stream: Stream): Promise<Buffer> => {
  return new Promise((resolve, reject) => {
    const chunks = [];
    stream.on('data', (chunk) => chunks.push(chunk));
    stream.on('end', () => resolve(Buffer.concat(chunks)));
    stream.on('error', (err) => reject(err));
  });
};
