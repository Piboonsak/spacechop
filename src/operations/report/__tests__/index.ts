import path from 'path';
import { PassThrough } from 'stream';
import streamToBuffer from '../../../lib/streamToBuffer';
import createTransformedStream from '../../../test/utils/createTransformedStream';
import toMatchImageSnapshot from '../../../test/utils/toMatchImageSnapshot';
import ImageDefinition from './../../../imagedef';
import { allFormats } from './../../../types/Format';
import { allFormats as allReportFormats } from './../../../types/ReportFormat';
import Report from './../index';
import { ReportConfig } from './../types';

expect.extend({ toMatchImageSnapshot });

describe('Report', () => {
  describe('Requirements', () => {
    const defaultConfig: ReportConfig = { format: 'json' };

    it('should not have any requirements', () => {
      const compress = new Report(defaultConfig);
      expect(compress.requirements()).toEqual([]);
    });
  });

  describe('Transformation of state', () => {
    const defaultConfig: ReportConfig = { format: 'json' };
    const defaultState: ImageDefinition = { height: 400, width: 400, type: 'jpeg' };

    it('should not return same state', () => {
      const r = new Report(defaultConfig);
      const stateBefore: ImageDefinition = { width: 100, height: 100, type: 'jpeg' };
      const { state: stateAfter } = r.execute(stateBefore);
      expect(stateBefore === stateAfter).toBe(false);
    });

    it('should not update width & height', () => {
      const op = new Report(defaultConfig);
      const { state } = op.execute(defaultState);
      expect(state).toEqual(expect.objectContaining({
        width: 400,
        height: 400,
      }));
    });

    it('should update mime', () => {
      const op = new Report(defaultConfig);
      const { state } = op.execute(defaultState);
      expect(state).toEqual(expect.objectContaining({
        mime: 'application/json',
      }));
    });
  });

  describe('Command', () => {
    const defaultConfig: ReportConfig = { format: 'json' };
    const defaultState: ImageDefinition = { height: 400, width: 400, type: 'jpeg'};

    it('should use echo', () => {
      const op = new Report(defaultConfig);
      const { command } = op.execute(defaultState);
      expect(command).toEqual(expect.stringMatching(/echo/));
    });
  });

  describe('Operation', () => {
    const defaultConfig: ReportConfig = { format: 'json' };
    const assetsFolder = path.join(__dirname, '../../../test/assets');
    const inputPaths = {
      jpeg: path.join(assetsFolder, 'grid.jpg'),
      png: path.join(assetsFolder, 'grid.png'),
      gif: path.join(assetsFolder, 'grid.gif'),
      webp: path.join(assetsFolder, 'grid.webp'),
    };

    // Test converting between different formats
    for (const fromFormat of allFormats) {
      for (const toFormat of allReportFormats) {
        describe(`report from [${fromFormat}] to [${toFormat}]`, async () => {

          const resultCopies = [];
          const numberOfCopies = 1;
          const source = inputPaths[fromFormat];
          const operation = new Report({ ...defaultConfig, format: toFormat });
          const state: ImageDefinition = { width: 100, height: 100, type: fromFormat };
          beforeAll(() => {
            const result = createTransformedStream(
              source,
              operation,
              state,
            );

            for (let i = 0; i < numberOfCopies; i++) {
              const pt = new PassThrough();
              result.pipe(pt);
              resultCopies.push(pt);
            }
          });

          it('should match snapshot', async () => {
            const buffer: Buffer = await streamToBuffer(resultCopies[0]);
            const data = JSON.parse(buffer.toString());
            expect(data).toEqual(
              expect.objectContaining(state),
            );
            await expect(data).toMatchSnapshot();
          });
        });
      }
    }
  });
});
