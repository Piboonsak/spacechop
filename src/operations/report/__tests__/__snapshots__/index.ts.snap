// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Report Operation report from [gif] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "gif",
  "width": 100,
}
`;

exports[`Report Operation report from [jpeg] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "jpeg",
  "width": 100,
}
`;

exports[`Report Operation report from [png] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "png",
  "width": 100,
}
`;

exports[`Report Operation report from [webp] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "webp",
  "width": 100,
}
`;

exports[`Report report from [gif] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "gif",
  "width": 100,
}
`;

exports[`Report report from [jpeg] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "jpeg",
  "width": 100,
}
`;

exports[`Report report from [png] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "png",
  "width": 100,
}
`;

exports[`Report report from [webp] to [json] should match snapshot 1`] = `
Object {
  "height": 100,
  "type": "webp",
  "width": 100,
}
`;
